src/treelearner/gpu_tree_learner.cpp

```diff
@@ -245,7 +245,7 @@ void GPUTreeLearner::AllocateGPUMemory() {
   }
   // allocate memory for all features (FIXME: 4 GB barrier on some devices, need to split to multiple buffers)
   device_features_.reset();
-  device_features_ = std::unique_ptr<boost::compute::vector<Feature4>>(new boost::compute::vector<Feature4>(num_dense_feature4_ * num_data_, ctx_));
+  device_features_ = std::unique_ptr<boost::compute::vector<Feature4>>(new boost::compute::vector<Feature4>((uint64_t)num_dense_feature4_ * num_data_, ctx_));
```

Comment:
@jiapengwen Thanks you for working on this! I think we should also have a fix here,
https://github.com/microsoft/LightGBM/blob/1d8bfd8c9b3fa7d56db81201adedb61cfc11669c/src/treelearner/gpu_tree_learner.cpp#L505-L506
could you help with this?

Reply:
yes, u r right, it may be overflow, I added for  robus